File: CS/DataAggregation/Form1.cs
using DevExpress.XtraCharts;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace DataAggregation {
    public partial class Form1 : DevExpress.XtraEditors.XtraForm {
        public Form1() {
            InitializeComponent();
        }

        List<DataPoint> GenerateData(int pointCount) {
            List<DataPoint> result = new List<DataPoint>(pointCount);
            double value = 0;
            DateTime argument = DateTime.Now.AddHours(-pointCount);
            Random random = new Random();
            
            for (double i = 0; i < pointCount; i++) {
                result.Add(new DataPoint { Argument = argument.AddHours(i), Value = Math.Abs(value) });
                value += (random.NextDouble() * 10.0 - 5.0);
            }
            return result;
        }

        #region #CustomAggregateFunction
        private void Form1_Load(object sender, EventArgs e) {
            Series series = chartControl.Series["Random Data"];
            series.DataSource = GenerateData(100_000);
            series.ArgumentDataMember = "Argument";
            series.ValueDataMembers.AddRange("Value", "Value", "Value", "Value");

            XYDiagram diagram = chartControl.Diagram as XYDiagram;
            diagram.AxisX.DateTimeScaleOptions.AggregateFunction = AggregateFunction.Custom;
            diagram.AxisX.DateTimeScaleOptions.CustomAggregateFunction = new OhlcAggregateFunction();
        }

        class OhlcAggregateFunction : CustomAggregateFunction {
            public override double[] Calculate(GroupInfo groupInfo) {
                double open = groupInfo.Values1.First();
                double close = groupInfo.Values1.Last();
                double high = Double.MinValue;
                double low = Double.MaxValue;
                foreach (double value in groupInfo.Values1) {
                    if (high < value) high = value;
                    if (low > value) low = value;
                }

                return new double[] { high, low, open, close };
            }
        }
        #endregion #CustomAggregateFunction

        class DataPoint {
            public DateTime Argument { get; set; }
            public double Value { get; set; }

        }
    }
}

File: VB/DataAggregation/Form1.vb
Imports DevExpress.XtraCharts
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace DataAggregation

    Public Partial Class Form1
        Inherits DevExpress.XtraEditors.XtraForm

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Function GenerateData(ByVal pointCount As Integer) As List(Of DataPoint)
            Dim result As List(Of DataPoint) = New List(Of DataPoint)(pointCount)
            Dim value As Double = 0
            Dim argument As DateTime = DateTime.Now.AddHours(-pointCount)
            Dim random As Random = New Random()
            For i As Double = 0 To pointCount - 1
                result.Add(New DataPoint With {.Argument = argument.AddHours(i), .Value = Math.Abs(value)})
                value +=(random.NextDouble() * 10.0 - 5.0)
            Next

            Return result
        End Function

#Region "#CustomAggregateFunction"
        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            Dim series As Series = chartControl.Series("Random Data")
            series.DataSource = Me.GenerateData(100_000)
            series.ArgumentDataMember = "Argument"
            series.ValueDataMembers.AddRange("Value", "Value", "Value", "Value")
            Dim diagram As XYDiagram = TryCast(chartControl.Diagram, XYDiagram)
            diagram.AxisX.DateTimeScaleOptions.AggregateFunction = AggregateFunction.Custom
            diagram.AxisX.DateTimeScaleOptions.CustomAggregateFunction = New OhlcAggregateFunction()
        End Sub

        Private Class OhlcAggregateFunction
            Inherits CustomAggregateFunction

            Public Overrides Function Calculate(ByVal groupInfo As GroupInfo) As Double()
                Dim open As Double = groupInfo.Values1.First()
                Dim close As Double = groupInfo.Values1.Last()
                Dim high As Double = [Double].MinValue
                Dim low As Double = [Double].MaxValue
                For Each value As Double In groupInfo.Values1
                    If high < value Then high = value
                    If low > value Then low = value
                Next

                Return New Double() {high, low, open, close}
            End Function
        End Class

#End Region  ' #CustomAggregateFunction
        Private Class DataPoint

            Public Property Argument As DateTime

            Public Property Value As Double
        End Class
    End Class
End Namespace
